cpp/src/groupby/sort/group_rank_scan.cu

```diff
@@ -93,53 +123,212 @@ std::unique_ptr<column> rank_generator(column_view const& order_by,
+  thrust::transform(rmm::exec_policy(stream),
+                    max_rank->view().begin<size_type>(),
+                    max_rank->view().end<size_type>(),
+                    min_rank->view().begin<size_type>(),
+                    mutable_ranks.begin<double>(),
+                    [] __device__(auto max_rank, auto min_rank) -> double {
+                      return min_rank + (max_rank - min_rank) / 2.0;
+                    });
+  return ranks;
+}
+
+std::unique_ptr<column> dense_rank_scan(column_view const& grouped_values,
+                                        column_view const& value_order,
                                         device_span<size_type const> group_labels,
                                         device_span<size_type const> group_offsets,
                                         rmm::cuda_stream_view stream,
                                         rmm::mr::device_memory_resource* mr)
 {
   return rank_generator(
-    order_by,
+    grouped_values,
+    value_order,
     group_labels,
     group_offsets,
     [] __device__(bool const unequal, size_type const) { return unequal ? 1 : 0; },
     DeviceSum{},
-    has_nested_nulls(table_view{{order_by}}),
+    has_nested_nulls(table_view{{grouped_values}}),
     stream,
     mr);
 }
 
-std::unique_ptr<column> percent_rank_scan(column_view const& order_by,
+std::unique_ptr<column> group_rank_to_percentage(bool is_dense_rank,
+                                                 column_view const& rank,
+                                                 column_view const& count,
+                                                 device_span<size_type const> group_labels,
+                                                 device_span<size_type const> group_offsets,
+                                                 rmm::cuda_stream_view stream,
+                                                 rmm::mr::device_memory_resource* mr)
+{
+  auto ranks = make_fixed_width_column(
+    data_type{type_to_id<double>()}, group_labels.size(), mask_state::UNALLOCATED, stream, mr);
+  ranks->set_null_mask(copy_bitmask(rank, stream, mr));
+  auto mutable_ranks = ranks->mutable_view();
+  thrust::tabulate(rmm::exec_policy(stream),
+                   mutable_ranks.begin<double>(),
+                   mutable_ranks.end<double>(),
+                   [is_dense_rank,
+                    is_double = rank.type().id() == type_id::FLOAT64,
+                    dcount    = count.begin<size_type>(),
+                    labels    = group_labels.data(),
+                    offsets   = group_offsets.begin(),
+                    d_rank    = rank.begin<double>(),
+                    s_rank    = rank.begin<size_type>()] __device__(size_type row_index) -> double {
+                     double const r   = is_double ? d_rank[row_index] : s_rank[row_index];
+                     auto const count = dcount[labels[row_index]];
+                     if (is_dense_rank) {
```

Comment:
Would it be viable to move the `if (is_dense_rank)` check out of the iteration? I wonder if we could check the value just once, and choose the appropriate lambda.